File: android/app/src/main/java/com/frenchnclc7/app/data/Plans.kt
package com.frenchnclc7.app.data

/** Premium billing periods. The price shown here is display text; what Google Play charges is set in the Play Console. */
enum class BillingPeriod(val amount: Int, val price: String, val per: String) {
    MONTHLY(12, "$12", "/ month"),
    YEARLY(120, "$120", "/ year"),
}

/** One cell of the Free-versus-Premium table. */
sealed interface PlanCell {
    data object Included : PlanCell
    data object NotIncluded : PlanCell
    data class Text(val value: String) : PlanCell
}

data class PlanRow(val label: String, val free: PlanCell, val premium: PlanCell)
data class PlanGroup(val title: String, val rows: List<PlanRow>)

/** The numbers and wording on the Plans page (the same as the web app's Plans page). */
object PlansLogic {
    /** Daily AI writing feedback per tier; the database table tier_limits enforces these. */
    const val AI_FEEDBACK_FREE = 1
    const val AI_FEEDBACK_PREMIUM = 5

    /** Yearly compared with paying monthly for a year. */
    val yearlySaving: Int = BillingPeriod.MONTHLY.amount * 12 - BillingPeriod.YEARLY.amount
    val yearlyPerMonth: Int = BillingPeriod.YEARLY.amount / 12
    val yearlyFreeMonths: Int = yearlySaving / BillingPeriod.MONTHLY.amount

    fun note(period: BillingPeriod): String = when (period) {
        BillingPeriod.YEARLY -> "$$yearlyPerMonth a month. $yearlyFreeMonths months free, save $$yearlySaving"
        BillingPeriod.MONTHLY -> "Billed monthly. Switch to yearly and save $$yearlySaving"
    }

    const val NOT_SWITCHED_ON = "Purchases aren't switched on yet. Please check back soon."
    const val PURCHASE_FAILED = "Couldn't start the purchase. Try again in a moment."

    val groups: List<PlanGroup> = listOf(
        PlanGroup("Study", listOf(
            PlanRow("Anki, Grammar, Kwiziq, TV5MONDE, Writing", PlanCell.Included, PlanCell.Included),
            PlanRow("Progress and streaks", PlanCell.Included, PlanCell.Included),
            PlanRow("Word Bank (your own words in Anki)", PlanCell.NotIncluded, PlanCell.Text("Up to ${WordBankLogic.PREMIUM_OWN_LIMIT} of your own")),
        )),
        PlanGroup("Lessons and links", listOf(
            PlanRow("Direct Kwiziq and TV5MONDE lesson links", PlanCell.Text("Search only"), PlanCell.Included),
        )),
        PlanGroup("Downloads and AI", listOf(
            PlanRow("Anki words per day", PlanCell.Text("${AnkiLimits.FREE_DAILY} per day"), PlanCell.Text("${AnkiLimits.PREMIUM_DAILY} per day")),
            PlanRow("Grammar chapter PDFs", PlanCell.NotIncluded, PlanCell.Included),
            PlanRow("Day-plan PDF", PlanCell.NotIncluded, PlanCell.Text("1 per day")),
            PlanRow("Writing AI feedback", PlanCell.Text("$AI_FEEDBACK_FREE per day"), PlanCell.Text("$AI_FEEDBACK_PREMIUM per day")),
        )),
    )
}

/** What happened when the person tried to buy or manage a subscription. */
sealed interface PurchaseResult {
    /** Google Play Billing isn't connected yet (no Play Console subscription products). */
    data object NotConfigured : PurchaseResult
    data object Failed : PurchaseResult
}

/**
 * The one place the app talks to Google Play Billing. Android requires Google's own billing for
 * subscriptions sold inside the app, so this is where the real implementation goes later
 * (BillingClient: query the two subscription products, launch the purchase flow, send the purchase
 * token to the server, which then sets the account's tier). Until then [NotConfiguredBilling] is used.
 */
interface PlayBilling {
    suspend fun purchase(period: BillingPeriod): PurchaseResult
    suspend fun manage(): PurchaseResult
}

/** Placeholder used until Google Play Billing is set up: nothing is charged and nothing is asked for. */
object NotConfiguredBilling : PlayBilling {
    override suspend fun purchase(period: BillingPeriod): PurchaseResult = PurchaseResult.NotConfigured
    override suspend fun manage(): PurchaseResult = PurchaseResult.NotConfigured
}

File: ios/FrenchNCLC7/Models/Plans.swift
import Foundation

/// Premium billing periods. The price shown here is display text; what Apple charges is set in App Store Connect.
enum BillingPeriod: CaseIterable, Equatable {
    case monthly, yearly

    var amount: Int { self == .monthly ? 12 : 120 }
    var price: String { self == .monthly ? "$12" : "$120" }
    var per: String { self == .monthly ? "/ month" : "/ year" }
    var label: String { self == .monthly ? "Monthly" : "Yearly" }
}

/// One cell of the Free-versus-Premium table.
enum PlanCell: Equatable {
    case included
    case notIncluded
    case text(String)
}

struct PlanRow: Equatable, Identifiable {
    let label: String
    let free: PlanCell
    let premium: PlanCell
    var id: String { label }
}

struct PlanGroup: Equatable, Identifiable {
    let title: String
    let rows: [PlanRow]
    var id: String { title }
}

/// The numbers and wording on the Plans page (the same as the web and Android apps).
enum PlansLogic {
    /// Daily AI writing feedback per tier; the database table tier_limits enforces these.
    static let aiFeedbackFree = 1
    static let aiFeedbackPremium = 5

    /// Yearly compared with paying monthly for a year.
    static let yearlySaving = BillingPeriod.monthly.amount * 12 - BillingPeriod.yearly.amount
    static let yearlyPerMonth = BillingPeriod.yearly.amount / 12
    static let yearlyFreeMonths = yearlySaving / BillingPeriod.monthly.amount

    static func note(_ period: BillingPeriod) -> String {
        switch period {
        case .yearly: return "$\(yearlyPerMonth) a month. \(yearlyFreeMonths) months free, save $\(yearlySaving)"
        case .monthly: return "Billed monthly. Switch to yearly and save $\(yearlySaving)"
        }
    }

    static let notSwitchedOn = "Purchases aren't switched on yet. Please check back soon."
    static let purchaseFailed = "Couldn't start the purchase. Try again in a moment."

    static let groups: [PlanGroup] = [
        PlanGroup(title: "Study", rows: [
            PlanRow(label: "Anki, Grammar, Kwiziq, TV5MONDE, Writing", free: .included, premium: .included),
            PlanRow(label: "Progress and streaks", free: .included, premium: .included),
            PlanRow(label: "Word Bank (your own words in Anki)", free: .notIncluded,
                    premium: .text("Up to \(WordBankLogic.premiumOwnLimit) of your own")),
        ]),
        PlanGroup(title: "Lessons and links", rows: [
            PlanRow(label: "Direct Kwiziq and TV5MONDE lesson links", free: .text("Search only"), premium: .included),
        ]),
        PlanGroup(title: "Downloads and AI", rows: [
            PlanRow(label: "Anki words per day", free: .text("\(AnkiLimits.freeDaily) per day"), premium: .text("\(AnkiLimits.premiumDaily) per day")),
            PlanRow(label: "Grammar chapter PDFs", free: .notIncluded, premium: .included),
            PlanRow(label: "Day-plan PDF", free: .notIncluded, premium: .text("1 per day")),
            PlanRow(label: "Writing AI feedback", free: .text("\(aiFeedbackFree) per day"), premium: .text("\(aiFeedbackPremium) per day")),
        ]),
    ]
}

/// What happened when the person tried to buy or manage a subscription.
enum PurchaseResult: Equatable {
    /// In-app purchases aren't connected yet (no App Store Connect subscription products).
    case notConfigured
    case failed
}

/// The one place the app talks to Apple's In-App Purchase (StoreKit). Apple requires its own billing for
/// subscriptions sold inside the app, so the real implementation goes here later (load the two subscription
/// products, run the purchase, send the receipt to the server, which then sets the account's tier).
/// Until then `NotConfiguredBilling` is used.
protocol AppStoreBilling {
    func purchase(_ period: BillingPeriod) async -> PurchaseResult
    func manage() async -> PurchaseResult
}

/// Placeholder used until In-App Purchase is set up: nothing is charged and nothing is asked for.
struct NotConfiguredBilling: AppStoreBilling {
    func purchase(_ period: BillingPeriod) async -> PurchaseResult { .notConfigured }
    func manage() async -> PurchaseResult { .notConfigured }
}
